applications/crossbar/src/modules/cb_rate_limits.erl

```diff
@@ -0,0 +1,231 @@
+%% @public
+%% @doc
+%% Does the path point to a valid resource
+%% So /rate_limits => []
+%%    /rate_limits/foo => [<<"foo">>]
+%%    /rate_limits/foo/bar => [<<"foo">>, <<"bar">>]
+%% @end
+%%--------------------------------------------------------------------
+-spec resource_exists() -> 'true'.
+resource_exists() -> 'true'.
+
+-spec only_if_reseller_or_master(cb_context:context()
+                                 ,fun((cb_context:context()) -> cb_context:context())
+                                ) -> cb_context:context().
+only_if_reseller_or_master(Context, Continue) ->
+    {'ok', MasterAccount} = whapps_util:get_master_account_id(),
+    AuthAccountId = cb_context:auth_account_id(Context),
+    case AuthAccountId =:= MasterAccount
+         orelse wh_services:is_reseller(AuthAccountId) of
+        'true' -> Continue(Context);
+        'false' -> cb_context:add_system_error('forbidden', Context)
+    end.
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%% Check the request (request body, query string params, path tokens, etc)
+%% and load necessary information.
+%% /rate_limits mights load a list of metaflow objects
+%% /rate_limits/123 might load the metaflow object 123
+%% Generally, use crossbar_doc to manipulate the cb_context{} record
+%% @end
+%%--------------------------------------------------------------------
+-spec validate(cb_context:context()) -> cb_context:context().
+validate(Context) ->
+    validate_rate_limits(Context, cb_context:req_verb(Context)).
+
+-spec validate_rate_limits(cb_context:context(), http_method()) -> cb_context:context().
+validate_rate_limits(Context, ?HTTP_GET) ->
+    validate_get_rate_limits(Context, thing_id(Context));
+validate_rate_limits(Context, ?HTTP_POST) ->
+    only_if_reseller_or_master(Context, fun validate_post_rate_limits/1);
+validate_rate_limits(Context, ?HTTP_DELETE) ->
+    only_if_reseller_or_master(Context, fun (C) -> validate_delete_rate_limits(C, thing_id(C)) end).
+
+-spec validate_post_rate_limits(cb_context:context()) -> cb_context:context().
+validate_post_rate_limits(Context) ->
+    case thing_type(Context) of
+        <<"accounts">> ->
+            cb_context:validate_request_data(<<"account_rate_limits">>, Context, fun validate_set_rate_limits/1);
+        <<"devices">> ->
+            cb_context:validate_request_data(<<"device_rate_limits">>, Context, fun validate_set_rate_limits/1);
+        _Else ->
+            crossbar_util:response_faulty_request(Context)
+    end.
+
+-spec thing_type(cb_context:context()) -> api_binary().
+thing_type(Context) ->
+    case cb_context:req_nouns(Context) of
+        [{<<"rate_limits">>, []}, {Type, [_ThingId]} | _] ->
+            Type;
+        _Nouns -> 'undefined'
+    end.
+
+-spec thing_id(cb_context:context()) -> api_binary().
+thing_id(Context) ->
+    case cb_context:req_nouns(Context) of
+        [{<<"rate_limits">>, []}, {<<"accounts">>, [AccountId]} | _] ->
```

Comment:
Why in `thing_type/1` do you use Type and _ThingId to match any preceding noun while here you are explicit about accounts/devices? I'd do something like:

```erlang
thing_type(Context) ->
    case thing_type_id(Context) of
        {Type, _Id} -> Type;
        'undefined' -> 'undefined'
end.

thing_id(Context) ->
    case thing_type_id(Context) of
        {_Type, Id} -> Id;
        'undefined' -> 'undefined'
    end.

thing_type_id(Context) ->
    case cb_context:req_nouns(Context) of
        [{<<"rate_limits">>, []}, {<<"accounts">> = Type, [Id]} | _] -> {Type, Id};
        [{<<"rate_limits">>, []}, {<<"devices">> = Type, [Id]} | _] -> {Type, Id};
        _ReqNouns -> 'undefined'
    end.
```
Now, if rate_limits gets extended to users or carriers or whatever else, its one place to change to permit this.